src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs

```diff
@@ -0,0 +1,196 @@
+            SnapshotPoint location, 
+            char typedChar, 
+            CancellationToken cancellationToken)
+            => s_commitChars.Contains(typedChar);
+
+        public AsyncCompletionData.CommitResult TryCommit(
+            IAsyncCompletionSession session, 
+            ITextBuffer subjectBuffer,
+            VSCompletionItem item, 
+            char typeChar, 
+            CancellationToken cancellationToken)
+        {
+            var document = subjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return CommitResultUnhandled;
+            }
+
+            var completionService = (CompletionServiceWithProviders)document.GetLanguageService<CompletionService>();
+            if (!item.Properties.TryGetProperty<RoslynCompletionItem>(Source.RoslynItem, out var roslynItem))
+            {
+                // This isn't an item we provided (e.g. Razor). Let the editor handle it normally.
+                return CommitResultUnhandled;
+            }
+
+            // We can be called before for ShouldCommitCompletion. However, that call does not provide rules applied for the completion item.
+            // Now we check for the commit charcter in the context of Rules that could change the list of commit characters.
+            if (!IsCommitCharacter(typeChar, roslynItem.Rules.CommitCharacterRules))
+            {
+                return new AsyncCompletionData.CommitResult(isHandled: true, AsyncCompletionData.CommitBehavior.CancelCommit);
+            }
+
+            if (completionService.GetProvider(roslynItem) is IFeaturesCustomCommitCompletionProvider featuresCustomCommitCompletionProvider)
+            {
+                // Custom commit provider assumes that null is provided is case of invoke. VS provides '\0' in the case.
+                char? commitChar = typeChar == '\0' ? null : (char?)typeChar;
+                var commitBehavior = CustomCommit(document, featuresCustomCommitCompletionProvider, session.TextView, 
+                    subjectBuffer, roslynItem, session.ApplicableToSpan, commitChar, cancellationToken);
+
+                return new AsyncCompletionData.CommitResult(isHandled: true, commitBehavior);
+            }
+
+            // TODO Remove language specific code: https://github.com/dotnet/roslyn/issues/30276
+            if (document.Project.Language == LanguageNames.VisualBasic && typeChar == '\n')
+            {
+                return new AsyncCompletionData.CommitResult(isHandled: false, 
+                    AsyncCompletionData.CommitBehavior.RaiseFurtherReturnKeyAndTabKeyCommandHandlers);
+            }
+
+            if (item.InsertText.EndsWith(":") && typeChar == ':')
+            {
+                return new AsyncCompletionData.CommitResult(isHandled: false, 
+                    AsyncCompletionData.CommitBehavior.SuppressFurtherTypeCharCommandHandlers);
+            }
```

Comment:
what is this code-block for?  This looks like this code is hardcoding in knowledge of a specific C#construct.  That is not acceptable and will not work properly for many other cases.  It will also potentially break other languages (like TS), which use ':', but don't necessarily want the same behavior as C#.  Please remove.

Reply:
It seems this should be resolved on provider level. I'll remove this in [https://github.com/dotnet/roslyn/issues/30276](https://github.com/dotnet/roslyn/issues/30276)

---
In reply to: [222520118](https://github.com/dotnet/roslyn/pull/29016#discussion_r222520118) [](ancestors = 222520118)